---
title: "Remake and adjust the figures in the manuscript"
output: pdf_document
---

### Load packages and initialize
```{r load_scripts, message=F, warning=F, echo=F, results='hide'}
#setwd("~/Dropbox (Harvard University)/analysis/analysis/2_figures/complete/compare_human")
setwd("~/Desktop/haber/ferret/ferret_ionocyte")
library(readr)
library(dplyr)
library(tidyr)
library(data.table)
library(tidyverse)
library(cowplot)
library(Seurat)
library(stringr)
library(data.table)
library(randomForest)
library(pheatmap)
library(pbapply)
source("r/util.r", chdir = T)
source("r/samples.r", chdir = T)
source("r/marker.r", chdir = T)
source("r/pca.r", chdir = T)
source("r/de_test.r", chdir = T)
theme_set(theme_cowplot()) # set the default theme for ggplot2]
frt = readRDS("~/Dropbox (Harvard University)/analysis/analysis/evan/combined_data/combined_corrected.rds")
source("r/clean_genes.R", chdir = T)

# Color maps
pal_use = "Rocket" # Heat Heat 2 Mako Rocket
fp_cols = rev(colorspace::sequential_hcl(palette = pal_use, n = 20)) #rev(colorspace::heat_hcl(20)) #weather_heat(25) # Feature Plot colors
ct_cols  = rainbow8equal #brewer16## cell type colors
channel_cols  = colorspace::qualitative_hcl(length(unique(frt$channel)), palette = "dynamic") ### channel/ferret colors
genotype_cols  = c("grey20", "red3")
```


```{r}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("GEOquery")

library(GEOquery)

gse <- getGEO("GSE103354", GSEMatrix=FALSE)

```

### Compare human, mouse and ferret rare cells particularly ionocytes
```{r Fig3_scRNA_overview}
#frt$channel = gsub("_kb", "", frt$orig.ident)
frt$type = "WT"
frt$type[frt$channel %in% c("K3", "D3", "X102", "Mix", "102", "Ionocyte1", "Ionocyte2", "Ionocyte3", "Ionocyte4")] = "WT-Enrich"
frt$type[grep("KO", frt$channel)] = "FOXI1-KO"
frt$type = factor(frt$type, levels=c("WT", "WT-Enrich", "FOXI1-KO"))
frt@meta.data$celltype[frt@meta.data$celltype == "Cycling Basal"] = "Basal"
frt@meta.data$celltype = factor(frt@meta.data$celltype, levels=rev(c("Ionocyte", "PNEC", "Tuft", "Basal", "Hillock", "Secretory", "Goblet", "Ciliated")))
frt@meta.data$genotype = factor(frt@meta.data$genotype, levels = c("WT", "FOXI1-KO"))

### Human, from lung cell atlas, sikemma et al, bioarxiv 2022
hs_cts = fread("~/Dropbox/HSPH/projects/nasal_mvc/analysis/191218_scRNASeq/adam/revision/human_lca/rare_cells/rarecells/X.csv")
hs_cts = t(hs_cts)
hs_genes = fread("~/Dropbox/HSPH/projects/nasal_mvc/analysis/191218_scRNASeq/adam/revision/human_lca/rare_cells/rarecells/var.csv")
hs_cells = fread("~/Dropbox/HSPH/projects/nasal_mvc/analysis/191218_scRNASeq/adam/revision/human_lca/rare_cells/rarecells/obs.csv")
rownames(hs_cts) = hs_genes$gene_symbols
colnames(hs_cts) = hs_cells$V1
hs = CreateSeuratObject(hs_cts)
hs@meta.data = cbind(hs@meta.data, hs_cells)
Idents(hs) = "ann_level_4"
vg = get.variable.genes.umis(GetAssayData(hs, slot="counts"), ret.plot = T, residual.threshold = -0.01, fit.spline = F)
VariableFeatures(hs) = vg$fit.data %>% arrange(residual) %>% head(n=2000) %>% dplyr::select(gene) %>% unlist() %>% unname() 
c("TRPM5", "ASCL3", "LTC4S", "PTGS1", "GNAT3", "ALOX5", "RGS1", "SH2D6", "BMX", "GFI1B", "CLDN10") %in% VariableFeatures(hs)
#iono = NormalizeData(object = iono,normalization.method = "LogNormalize", scale.factor = 10000) ### log2(TPM+1) normalization
hs = ScaleData(object = hs, verbose = TRUE) ### scale data
hs = RunPCA(hs, features = VariableFeatures(hs)) 
hs = RunUMAP(hs, reduction = "pca", dims = 1:5, min.dist = 0.75, n.neighbors = 25) 
hs = FindNeighbors(hs, reduction = "pca", dims = 1:5, nn.eps = 0)
hs = FindClusters(object = hs, resolution = 0.25, verbose = TRUE)
hs$celltype = hs$ann_level_4


### Mouse, from montoro et al, pulse-seq
mm = readRDS("~/Dropbox/Postdoc/Projects/airway_injury/HDM/data/pulse_seq/ps_complete_V3.rds")
mm_rare = subset(mm, res.2_named %in% c("Ionocyte", "Neuroendocrine", "Tuft"))
ggsave(ViolinP(subset(hs, celltype=="Ionocyte"), c("VMO1", "LYPD2", "CD74", "ANXA3", "PPP1R1A", "SLC12A2"), group.by = "anatomical_region_level_2", legend = T) + scale_fill_manual("", values=brewer.pal(9, "Set1")), filename="iono_subtype_markers_region_lev2.pdf")
ggsave(ViolinP(subset(hs, celltype=="Ionocyte"), c("VMO1", "LYPD2", "CD74", "ANXA3", "PPP1R1A", "SLC12A2"), group.by = "anatomical_region_level_1", legend = T)+ scale_fill_manual("", values=brewer.pal(3, "Set2")), filename="iono_subtype_markers_region_lev1.pdf")

```
Integrate the data from the 3 species and save
```{r overall_compare_rarecells}
### Overall similarity between human and ferret rare cells.
frt_rare = subset(frt, celltype %in% c("Tuft", "PNEC", "Ionocyte"))
frt_rare$species = "ferret"
frt_rare$celltype = gsub("PNEC", "Neuroendocrine", frt_rare$celltype)
hs$species = "human"
mg = convert_mouse_to_human(rownames(mm_rare))
mm_rare_cts = GetAssayData(mm_rare, slot="counts")
mm_rare_cts = mm_rare_cts[which(!is.na(mg)),]
rownames(mm_rare_cts) = mg[!is.na(mg)]
mm_rare_gc = CreateSeuratObject(mm_rare_cts) # gc - genes converted
mm_rare_gc = NormalizeData(mm_rare_gc)
mm_rare_gc$species = "mouse"
mm_rare_gc$celltype = mm_rare$res.2_named

rareall = merge(frt_rare, hs) 
rareall = merge(rareall, mm_rare_gc)
### visualize the gene name overlap.
VennDiagram::venn.diagram(
    x = list(rownames(hs), rownames(frt_rare), rownames(mm_rare_gc)),
    category.names = c("Human" , "Ferret" , "Mouse"),
    filename = 'gene_name_overlap.png',
    output=TRUE
)

inboth = intersect(clean_genes(rownames(frt)), rownames(hs))
inboth = intersect(inboth, rownames(mm_rare_gc))

# use one-way anova to identify species-specific genes
d = FetchData(rareall, c(inboth, "species"))
colnames(d) = make.names(colnames(d)) #"NKX3-2" "NKX3-1" wont work as formulae
cl = parallel::makeCluster(8)
parallel::clusterExport(cl=cl, list("d"),  envir=environment()) # this is memory intensive but there's not that many cells so it works.
f = pblapply(setdiff(colnames(d), "species"), function(x) as.formula(sprintf("%s ~ species", x)))
res = pblapply(f, function(x) summary(aov(x, data = d)), cl = cl) # parallelization isn't necessary but it saves some time. 4 min -> 30 sec
pval = unlist(pblapply(res, function(x) x[[1]]$"Pr(>F)"[1]))
fstat = unlist(pblapply(res, function(x) x[[1]]$"F value"[1]))
anova_output = data.frame(gene = setdiff(colnames(d), "species"), pval=pval, fstat=fstat)
anova_output = subset(anova_output, !is.na(fstat))
anova_output$species_specific = anova_output$fstat < quantile(anova_output$fstat, 0.9, na.rm = T)
write.table(anova_output, file="species_specific_anova_results.txt", sep = "\t", quote=F, row.names = F)
genes_use = anova_output$gene[anova_output$species_specific]
genes_use = gsub(".", "-", genes_use, fixed=T) # watch out for e.g NKX2.2 / NKX2-2
vg = get.variable.genes.umis(GetAssayData(rareall, slot="counts")[genes_use,], ret.plot = T, residual.threshold = -0.01, fit.spline = F)
VariableFeatures(rareall) = vg$fit.data %>% arrange(residual) %>% head(n=2000) %>% dplyr::select(gene) %>% unlist() %>% unname() 
c("TRPM5", "ASCL3", "LTC4S", "PTGS1", "GNAT3", "ALOX5", "RGS1", "SH2D6", "BMX", "GFI1B", "CLDN10") %in% VariableFeatures(rareall)
#iono = NormalizeData(object = iono,normalization.method = "LogNormalize", scale.factor = 10000) ### log2(TPM+1) normalization
rareall = ScaleData(object = rareall, verbose = TRUE) ### scale data
run_batch_correct = T
if(run_batch_correct){
    obj_list = SplitObject(rareall, split.by = "species")
    features = SelectIntegrationFeatures(object.list = obj_list)
    anchors = FindIntegrationAnchors(object.list = obj_list, anchor.features = features, dims = 1:10)
    rareall = IntegrateData(anchorset = anchors, dims = 1:10) #https://github.com/satijalab/seurat/issues/3930
    DefaultAssay(rareall) = "integrated"
    rareall = ScaleData(rareall, verbose = FALSE)
    rareall = RunPCA(rareall, features = VariableFeatures(rareall)) 
    #rareall = RunPCA(rareall, features = VariableFeatures(rareall)) 
    #rareall = rareall %>% harmony::RunHarmony("species", project.dim=F)
}

### UMAP
rareall = RunUMAP(rareall, reduction = "pca", dims = 1:10, min.dist = 0.25, n.neighbors = 100) 
rareall$celltype_species = paste(rareall$celltype, rareall$species, sep=" - ")
saveRDS(rareall, file="rareall.rds")

```

# Compare the rare cells of each type, particularly ionocytes
```{r comparison_analysis}
rareall = readRDS("rareall.rds")
ggsave(DimPlot(rareall, group.by = "celltype", reduction = "umap") + scale_color_manual("", values = brewer.pal(3, "Set1")), filename = "umap_celltype.pdf", width=7, height=6)
ggsave(DimPlot(rareall, group.by = "celltype_species", reduction = "umap"), filename = "umap_celltype_species.pdf", width=7, height=6)
ggsave(DimPlot(rareall, group.by = "species", reduction = "umap"), filename = "umap_species.pdf", width=7, height=6)

### Correlation matrix
cm = cor(t(FetchData(rareall, VariableFeatures(rareall), slot = "data")))
cm = data.frame(cm)
cm$celltype = rareall$celltype
cm$species = rareall$species
cmg = cm %>% group_by(celltype, species) %>% summarize_all(.funs=mean)
cmg = cmg %>% pivot_longer(cols = -c(celltype, species), names_to = "cell", values_to = "mean_cor")
cmg$from_species = rareall$species[match(cmg$cell, colnames(rareall))]
cmg$from_celltype = rareall$celltype[match(cmg$cell, colnames(rareall))]
# Compute summary statistics for correlations:
d = cmg %>% group_by(celltype, species, from_species, from_celltype) %>% 
    dplyr::summarise(boxplot = list( setNames(boxplot.stats(mean_cor)$stats,                                  c('lower_whisker','lower_hinge','median','upper_hinge','upper_whisker') ) ) ) %>%
    unnest_wider(boxplot)

# grab the nine comparison p-values we need to label the plot:
compars = data.frame(x = c("human-Ionocyte-human-Ionocyte", "mouse-Ionocyte-human-Ionocyte", "mouse-Ionocyte-human-Ionocyte", "human-Neuroendocrine-human-Neuroendocrine", "mouse-Neuroendocrine-human-Neuroendocrine", "mouse-Neuroendocrine-human-Neuroendocrine", "human-Tuft-human-Tuft", "mouse-Tuft-human-Tuft", "mouse-Tuft-human-Tuft"), y= c( "ferret-Ionocyte-human-Ionocyte", "human-Ionocyte-human-Ionocyte", "ferret-Ionocyte-human-Ionocyte", "ferret-Neuroendocrine-human-Neuroendocrine", "human-Neuroendocrine-human-Neuroendocrine", "ferret-Neuroendocrine-human-Neuroendocrine", "ferret-Tuft-human-Tuft", "human-Tuft-human-Tuft", "ferret-Tuft-human-Tuft"))
cmg = mutate(cmg, group = paste(species, celltype, from_species, from_celltype, sep="-"))
m = pairwise.wilcox.test(cmg$mean_cor, g = cmg$group)$p.value
compars$pval = apply(compars, 1, function(row){m[row[1], row[2]]})
compars$sig  = gtools::stars.pval(compars$pval)
write.table(compars, file = "wilcoxon_pvals_correlation_comparison.csv", sep = ",", row.names = F, quote=F)

d$species = factor(d$species, levels=c("human", "ferret", "mouse"))
ggsave(ggplot(subset(d, from_species=="human"), aes(x=species, fill=celltype, ymin = lower_whisker, lower = lower_hinge, middle = median, upper = upper_hinge, ymax = upper_whisker))  + geom_boxplot(stat = "identity") + facet_wrap(~from_celltype) + ylab("Mean correlation with Human") + scale_fill_manual("", values = brewer.pal(3, "Set1")) + xlab(""), filename="mean_cor_by_species.pdf", width=8, height=4)

# downsample and visualize the correlation matrix
subsample = 250
use_cells = resample(rareall$celltype, max.group.size = subsample, return.index = T)
cell_order = c("Ionocyte", "Neuroendocrine", "Tuft")
rcs = subset(rareall, cells=colnames(rareall)[use_cells])
md = rcs@meta.data[, c("nFeature_RNA","celltype", "species")] # type
md$celltype = factor(rcs$celltype, levels=cell_order)
ct_cols  = rainbow8equal #brewer16## cell type colors
channel_cols  = colorspace::qualitative_hcl(length(unique(frt$channel)), palette = "dynamic") ### channel/ferret colors
species_cols  = brewer.pal(3, "Set1")
names(ct_cols) = cell_order
names(channel_cols) = unique(frt$channel)
names(species_cols)= unique(rcs$species)
column_colors = list("channel"=channel_cols, "species"=species_cols, "celltype"=ct_cols[1:length(cell_order)])
cm = cor(t(FetchData(rcs, paste0("PC_", 1:10))))
dc = diverging_colors(cm, n=20, brewer_col_range = 7)
pheatmap(cm, show_colnames = F, show_rownames = F, cluster_rows = F, cluster_cols = F, annotation_col = md, annotation_colors = column_colors, color = dc$colors, breaks = dc$breaks, filename="hs_frt_cormat.png", width=10, height=9)



### Differences. Core markers and DE by celltype.
#species_specific = spc@all$GENE_SYMBOL[abs(spc@all$alpha.diff) > 0.1]
#de_iono = compare.groups(GetAssayData(subset(rareall, celltype=="Ionocyte")), groups=subset(rareall, #celltype=="Ionocyte")$species, n.cores = 12)
#de_iono = de_iono@all %>% arrange(-abs(log2fc))
#ggsave(ggplot(subset(de_iono, !GENE_SYMBOL %in% species_specific), aes(x=-log2fc, y=-log10(p.adj), label=GENE_SYMBOL)) + #geom_point() + ggrepel::geom_text_repel(color="red4", data=subset(de_iono, !GENE_SYMBOL %in% species_specific & log2fc > 0.5)) + 
#ggrepel::geom_text_repel(color="blue4", data=subset(de_iono, !GENE_SYMBOL %in% species_specific & log2fc < -0.5)), filename="iono_volcano.pdf")

source("~/hldev/rna_seq/r/marker_compare.r")
ionchans = read.delim("~/Dropbox/HSPH/projects/ferret_ionocyte/hsph/analysis/evan/ionchannelsubset1.csv", sep = ",")$Genes
iono = subset(rareall, celltype=="Ionocyte")
tpm = GetAssayData(iono, assay = "RNA")
m = FetchData(iono, c(paste0("rna_", ionchans), "species"))
m = m |> group_by(species) |> summarize_all(.funs = mean) |> data.frame()
colnames(m) = gsub("rna_", "", colnames(m))
rownames(m) = m$species
m$species = NULL
m = t(m)
m = m[order(rownames(m)),]
write.table(m, file = "all_ionochannels_in_ionos_by_species.csv", sep = ",", quote=F)


use.genes = unique(c(VariableFeatures(iono), rownames(umi[Matrix::rowSums(umi > 0) > 5,])))
m = markers.compare(norm.counts = tpm, cluster.labels = iono$species, test="mast",  n.threads = 12, use.genes = use.genes)
z= markers.rank(compare.obj = m, counts = tpm, clusters = iono$species, max.Q=0.05, min.fc=0.5, fc.metric="FC.Min", Q.metric="Q.Max", fc.col="log2fc", min.frac.express=0)

mk = read.delim("top_500_filtered_FC.Min_0.5_Q.Max_0.05_frac.expr_0.txt")
mk = prune.df.with.list(mk, ionchans)
mk = mk[, c("human", "ferret", "mouse")]
genes = melt(mk, id.vars = NULL)[,2]
genes = genes[!is.na(genes)]
iono$species = factor(iono$species, levels=c("human", "ferret", "mouse"))
ggsave(DotPlot(iono, features = c("CFTR", "ASCL3", "FOXI1", "BSND", rev(genes)), group.by = "species", scale = F, assay = "RNA") + coord_flip() + xlab("") + ylab("") + scale_color_gradientn(colours =  fp_cols) + scale_radius(range = c(0,6.5)), filename="iono_species_channels.pdf", width=5, height=8)



ggsave(DotPlot(iono, features = rev(c("ATP10D", "ATP12A","ATP1B1","ATP6V0A4","ATP6V0C","ATP6V1G3","BSND","CFTR","CLCNKA","FXYD2","KCNK1","KCNQ1","SCNN1A","SCNN1B","SCNN1G","SLC12A2","SLC16A3",
"SLC25A6")), group.by = "species", scale = F, assay = "RNA") + coord_flip() + xlab("") + ylab("") + scale_color_gradientn(colours =  fp_cols) + scale_radius(range = c(0,12)), filename="iono_species_channels_GRC.pdf", width=5, height=8)

```

### Markers of potential common progenitor.
```{r }
rare = readRDS("rareall.rds")
rare = FindNeighbors(rare, reduction = "pca", dims = 1:10) 
rare = FindClusters(rare, resolution = 0.25) 
m = FindMarkers(rare, ident.1 = 4)
rare$isprog = "Other"
rare$isprog[rare$integrated_snn_res.0.25==4] = "Progenitor"

d = GetAssayData(rare, assay = "RNA")
nzcount = Matrix::rowSums(d > 1)
test_genes= names(nzcount[nzcount > 10])
#de = compare.groups(d = d[test_genes,], compare = "Progenitor", groups = rare$isprog, test = "mast", n.cores = 1, batch= rare$channel)
#write.table(de@all, file = "de_compare_prog_vs_bkgd.txt", sep = "\t", quote=F, row.names = F)
de = read.delim("de_compare_prog_vs_bkgd.txt")

mplot = de
mplot$gene = mplot$GENE_SYMBOL
mplot$logfdr = -log10(mplot$p.adj)
mplot$logfdr[mplot$logfdr > 100] = 100
ggsave(ggplot(mplot, aes(x=log2fc, y=logfdr, label=gene)) + geom_point(color="black") + geom_point(data=subset(mplot, log2fc > 1), color="red") + geom_point(data=subset(mplot, log2fc < -0.5), color="green3") + geom_text_repel(data=subset(mplot, log2fc > 1), color="red") + geom_text_repel(data=subset(mplot, log2fc < -0.75), color="green3") + xlab("Progenitor vs others, log2fc") + theme_cowplot(), filename="prog_de.pdf")


ng = 50
maxz = 2
genes_use = filter(read.delim("species_specific_anova_results.txt"), species_specific) %>% select(gene) %>% unlist() %>% unname()
top = de %>% filter(p.adj < 0.01 & GENE_SYMBOL %in% genes_use) %>% arrange(-log2fc) %>% head(n=ng) %>% dplyr::select(GENE_SYMBOL) %>% unlist()
bottom = de %>% filter(p.adj < 0.01  & GENE_SYMBOL %in% genes_use) %>% arrange(log2fc) %>% head(n=ng) %>% dplyr::select(GENE_SYMBOL) %>% unlist()
use = resample(rare$isprog, max.group.size = 500, return.index = T)
md = rare@meta.data[use, c("isprog", "species", "nFeature_RNA")]
z = d[c(top, bottom),use]
z = t(scale(t(z)))
z[z > maxz] = maxz
z[z < -maxz] = -maxz
dc = diverging_colors(z, n=20, brewer_col_range = 8)

# dont have any ordering by species. so first shuffle
ix = sample(1:nrow(md))
md = md[ix,]
z = z[,ix]
# then order by whether or not a progenitor
ix = order(md$isprog)
md = md[ix,]
z = z[,ix]

ct_cols = c("grey90", "grey10")
names(ct_cols) = unique(md$isprog)
column_colors = list("isprog"=ct_cols)

pheatmap(z, show_colnames = F, cluster_cols = F, clustering_method  = "ward.D2", clustering_distance_rows = "correlation", filename = "prog_de_heatmap.pdf", color = dc$colors, annotation_col = md, annotation_colors = column_colors, breaks = dc$breaks, width = 14, height=12)

# write the table of progenitor-enriched genes
m = subset(de, GENE_SYMBOL %in% genes_use) %>% arrange(-log2fc)
m = m[, c("GENE_SYMBOL", "p", "p.adj", "log2fc", "log2fc_MAST")]
colnames(m) = c("gene", "p-value", "FDR", "log2fc mean", "log2fc MAST")
write.table(m, file = "de_compare_prog_vs_bkgd_nospecies_specific.txt", sep = "\t", quote=F, row.names=F)

```


### Subtypes. Are ferret subtypes conserved in human.
```{r subtypes}
hs_iono = subset(hs, celltype=="Ionocyte")
vg = get.variable.genes.umis(GetAssayData(hs_iono, slot="counts"), ret.plot = T, residual.threshold = -0.01, fit.spline = F)
VariableFeatures(hs_iono) = vg$fit.data %>% arrange(residual) %>% head(n=2000) %>% dplyr::select(gene) %>% unlist() %>% unname() 
hs_iono = ScaleData(object = hs_iono, verbose = TRUE) ### scale data
hs_iono = RunPCA(hs_iono, features = VariableFeatures(hs)) 
hs_iono = RunUMAP(hs_iono, reduction = "pca", dims = 1:5, min.dist = 0.75, n.neighbors = 25) 
hs_iono = FindNeighbors(hs_iono, reduction = "pca", dims = 1:5, nn.eps = 0)
hs_iono = FindClusters(object = hs_iono, resolution = 0.1, verbose = TRUE)
frt_iono_markers = read.delim("../ionocyte/markers/top_500_filtered_FC.Min_0.1_Q.Max_0.05_frac.expr_0.txt")
cell_cycle = read.delim("~/ref/gene_lists/gut_circuits/signatures/Gene.Lists.FINAL.Atlas.MHC.papers.txt", stringsAsFactors = F)$Cell.Cycle
cell_cycle = sample(intersect(toupper(cell_cycle), rownames(hs_iono)), 20)
hs_iono$cell_cycle = Matrix::colMeans(GetAssayData(hs_iono)[cell_cycle,])
hs_iono$ferret_iono_typeA = Matrix::colMeans(GetAssayData(hs_iono[frt_iono_markers$Type_A[1:20],]))
hs_iono$ferret_iono_typeB = Matrix::colMeans(GetAssayData(hs_iono[frt_iono_markers$Type_B[1:20],]))
hs_iono$ferret_iono_typeC = Matrix::colMeans(GetAssayData(hs_iono[frt_iono_markers$Type_C[1:20],]))
ggsave(FeatureP(hs_iono, c("ferret_iono_typeA", "ferret_iono_typeB", "ferret_iono_typeC"), pt.size = 2, light.theme = T, ncol = 3) , filename="ferret_iono_subtype_sigs_feature.pdf", width=10, height=3)

### Sort of looks like the 3 ferret subtypes might be contained in clusters 1 and 2. what's cluster 0?
x = compare.groups(GetAssayData(hs_iono), groups = hs_iono$RNA_snn_res.0.1, compare=0, n.cores = 12)
x = x@all %>% arrange(-abs(log2fc))
ggsave(FeatureP(hs_iono, c("S100A6", "CFTR", "DEFB1", "CD81", "S100A4", "ATP6V1G3", "KRT7", "WFDC2", "KRT19"), pt.size = 2, light.theme = T, ncol = 3) , filename="ferret_iono_clust0_feature.pdf", width=10, height=10)

```


### Heterogenity in human ionocytes figures
```{r unsup_human_iono_heterog}
markers = read.delim("../../../evan/combined_data/markers/celltype/top_500_filtered_FC.Min_0.1_Q.Max_0.05_frac.expr_0.txt")
iono$secretory = Matrix::colMeans(GetAssayData(iono[markers$Secretory[1:20],]))
hs_iono$basal = Matrix::colMeans(GetAssayData(hs_iono[markers$Basal[1:20],]))

md = FetchData(hs_iono, c("nFeature_RNA","secretory", "basal", "channel", "PC_1", "PC_2", "PC_3")) # type
#names(ct_cols) = cell_order
names(type_cols) = unique(md$type)
pc1_cols = colorspace::diverging_hcl(palette = "Green-Orange", n=10)
pc2_cols = colorspace::diverging_hcl(palette = "Tropic", n=10)
pc3_cols = colorspace::diverging_hcl(palette = "Purple-Brown", n=10)
type_cols = brewer16[1:4]
channel_cols  = colorspace::qualitative_hcl(length(unique(frt$channel)), palette = "dynamic") ### channel/ferret colors
names(channel_cols) = sort(unique(md$channel))
column_colors = list("channel"=channel_cols[1:length(unique(md$channel))], "PC_1"=pc1_cols, "PC_2"=pc2_cols, "PC_3"=pc3_cols)
cm = cor(t(FetchData(iono, paste0("PC_", 1:20))))
dc = diverging_colors(cm, n=20, brewer_col_range = 9)
pheatmap(cm, show_colnames = F, show_rownames = F, clustering_method = "ward.D2", clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", annotation_col = md, annotation_colors = column_colors, color = dc$colors, breaks = dc$breaks, filename="iono_cormat.png", width=12, height=11)

ggsave(FeaturePlot(iono, "PC_1") + scale_color_gradientn(colors = pc1_cols), filename="umap_PC1.pdf", width=5, height=4.4)
ggsave(FeaturePlot(iono, "PC_2") + scale_color_gradientn(colors = pc2_cols), filename="umap_PC2.pdf", width=5, height=4.4)
ggsave(FeaturePlot(iono, "PC_3") + scale_color_gradientn(colors = pc3_cols), filename="umap_PC3.pdf", width=5, height=4.4)
ggsave(DimPlot(iono, label=F) + scale_color_manual(values = type_cols) + xlab("UMAP-1") + ylab("UMAP-2"), filename = "umap_cluster.pdf", width=5.5, height=4.4)
```


### Show eGFP expression in the progenitor. Is it positive?
```{r }
rareall = readRDS("rareall.rds")


gfp = readRDS("~/Dropbox/HSPH/projects/ferret_ionocyte/hsph/analysis/1_initial_cluster/align_2021genome_kallisto_GFP_tdTom/ferret_2021_kallisto_gfpTdTom_processed.rds")
gfp$celltype = forcats::fct_collapse(gfp$RNA_snn_res.0.1, Ionocyte="10", PNEC="9", Tuft="11", Serous="12", Basal=c("4", "5", "6"), Secretory=c("0", "1", "3", "6", "7", "8"), Ciliated=c("2", "7"))
gfp = SetIdent(gfp, value="celltype")
gfp$channel = gfp$barcode # have to swap annotations here because the GFP object is misannotated
gfp$channel[grep("KO_1", colnames(gfp))] = "KO_1_M"
gfp$channel[grep("KO_2", colnames(gfp))] = "KO_2_F"
gfp$channel[grep("KO_3", colnames(gfp))] = "KO_3_F"
gfp$channel[grep("KO_4", colnames(gfp))] = "KO_4_M"
gfp$channel = gsub("iono", "Ionocyte", gfp$channel)
longest_substr = function(x, delim="_"){y = unlist(str_split_fixed(x, delim, 1+str_count(x, delim))); y[which.max(nchar(y))]} #adam's solution
longest_substr = function(y) sapply(strsplit(y, "_"), function(x) x[which.max(nchar(x))]) # stack overflow
gfp$barcode = unlist(lapply(colnames(gfp), longest_substr)) # the barcode column didn't actually have barcodes, this fixes it.
frt$barcode = unlist(lapply(colnames(frt), longest_substr))
gfp$cell_id = paste(gfp$channel, gfp$barcode, sep="_")

md = FetchData(gfp, c("eGFP", "tdTomato"), slot="count")
colnames(md) = paste0(colnames(md), "_UMIs")
md = cbind(md, FetchData(gfp, c("eGFP", "tdTomato", "cell_id")))

rareall$cell_id = colnames(rareall)
md = merge(rareall@meta.data, md, by = "cell_id", all.x = T)
rownames(md) = md$cell_id
rareall@meta.data = md

pal_use = "Rocket" # Heat Heat 2 Mako Rocket
fp_cols = rev(colorspace::sequential_hcl(palette = pal_use, n = 20))
rareall@meta.data$eGFP[is.na(rareall@meta.data$eGFP)] = 0
rareall@meta.data$eGFP_UMIs[is.na(rareall@meta.data$eGFP_UMIs)] = 0
FeaturePlot(rareall, "eGFP", cols = fp_cols, pt.size = 0.5)

x = FetchData(rareall, c("eGFP", "TRPM5", "CHGA", "BSND", "FOXI1", "UMAP_1", "UMAP_2")) %>% pivot_longer(cols=-c(UMAP_1, UMAP_2))
ggplot(x, aes(x=UMAP_1, y=UMAP_2, color=value)) + geom_point() + facet_wrap(~name) + scale_color_gradientn(colors=fp_cols)

x = FetchData(rareall, c("eGFP", paste("rna", c("TRPM5", "CHGA", "BSND", "FOXI1"), sep="_"), "UMAP_1", "UMAP_2"))
colnames(x) = gsub("rna_", "", colnames(x))
ggsave(cowplot::plot_grid(plotlist=lapply(c("TRPM5", "CHGA", "FOXI1", "eGFP"), function(y) ggplot(x, aes_string(x="UMAP_1", y="UMAP_2", color=y)) + geom_point(stroke=0, size=0.75) + scale_color_gradientn(colors=fp_cols))),filename="rarecells_multispecies_markers.pdf", width=8.5, height=7)
ggsave(cowplot::plot_grid(plotlist=lapply(c("TRPM5", "CHGA", "BSND", "FOXI1", "eGFP"), function(y) ggplot(x, aes_string(x="UMAP_1", y="UMAP_2", color=y)) + geom_point(stroke=0, size=0.75) + scale_color_gradientn(colors=fp_cols))),filename="rarecells_multispecies_markers_BSND.pdf", width=13, height=7)
ggsave(cowplot::plot_grid(plotlist=lapply(c("TRPM5", "CHGA", "BSND", "FOXI1"), function(y) ggplot(x, aes_string(x="UMAP_1", y="UMAP_2", color=y)) + geom_point(stroke=0, size=0.75) + scale_color_gradientn(colors=fp_cols))),filename="rarecells_multispecies_markers_noEGFP.pdf", width=8.5, height=7)

### TODO:
### is pulse-seq data labeled earlier
mm_basal = subset(mm, res.2_named == "Basal")
str_split_fixed(colnames(subset(rareall, species=="mouse")), "_", 4)[,4]
```

